### PR TITLE
Add WebAuthn 2FA in UI

### PR DESCRIPTION
**Pull request by grzuy**
First stab at #1948 in collaboration with @padulafacundo.

Prototype adding experimental support for WebAuthn as a 2nd factor.

Goal of this first prototype is to have a working solution that can be tested by rubygems.org and community developers to get early feedback and start the discussion.

### What it does

- Let's a user who already has 2FA (OTP) configure register WebAuthn Security Keys
- On UI Sign In, user can 2FA with WebAuthn Security or fallback to OTP

#### Pending:

- [x] Ability to delete the registered WebAuthn credential
- [x] Support multiple WebAuthn credentials
- [x] Check the User Agent supports WebAuthn
- [x] Better handle Sign In failures
- [x] user id randomness in credential creation options (https://www.w3.org/TR/webauthn-2/#user-handle)
- [x] Fix 3 failing tests
- [x] Polish credential list UI
- [x] Polish WebAuthn session prompt UX
- [x] Adjust user-facing terminology ("Credential" vs. "Security Key")
- [x] Integration with OTP setup in the "Edit profile" page
- [x] Show "Last used" for each security key to the user 
- [x] Update sign_count to bigint
- [x] Rate limit WebAuthn routes
- [x] Push all encoding/decoding to the [`webauthn`](https://github.com/cedarcode/webauthn-ruby) gem
- [x] Fix CI build docker step
- [x] Let user know what's happening when registration of same security key fails
- [x] Remove encoding in webauthn tests
- [x] Depend on a `webauthn` gem released version, not `master` (after gem release)

Follow up items:
- Support WebAuthn 2FA at the API level

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -25,6 +25,8 @@ class User < ApplicationRecord
   has_many :deletions, dependent: :nullify
   has_many :web_hooks, dependent: :destroy
 
+  has_many :credentials, dependent: :destroy
```
Comment (grzuy) [inline app/models/user.rb]: Probably less confusing if we change this to `webauthn_credentials`?

File: app/views/layouts/application.html.erb [hunk 2]
```diff
@@ -223,5 +223,6 @@
     <%= yield :javascript %>
     <script type="text/javascript" async defer id="gauges-tracker" data-site-id="4eab0ac8613f5d1583000005" src="//secure.gaug.es/track.js"></script>
     <script type="text/javascript" defer src="https://www.fastly-insights.com/insights.js?k=3e63c3cd-fc37-4b19-80b9-65ce64af060a"></script>
+    <script type="text/javascript" src="https://unpkg.com/@github/webauthn-json@0.3.1/dist/index.js"></script>
```
Comment (grzuy) [inline app/views/layouts/application.html.erb]: Is this OK or should be use `vendor/assets/javascripts`?

File: app/views/sessions/webauthn_prompt.html.erb [hunk 3]
```diff
@@ -0,0 +1,10 @@
+<% @title = t('webauthn_authentication') %>
+
+<%= form_tag webauthn_authentication_options_session_path, method: :post, id: 'webauthn-sign-in' do %>
+  <div class="text_field">
+    <%= label_tag :webauthn, t('webauthn_credentials.sign_in_prompt'), class: 'form__label' %>
+  </div>
+  <div class="form_bottom">
+    <%= submit_tag t('sign_in'), data: {disable_with: t('form_disable_with')}, class: 'form__submit' %>
+  </div>
```
Comment (grzuy) [inline app/views/sessions/webauthn_prompt.html.erb]: Will `button_to` be enough? As opposed to the `form_tag`?

File: app/views/layouts/application.html.erb (hunk 2 above)
Comment (DanielHeath) [reply app/views/layouts/application.html.erb]: This grants anyone with deploy access at `unpkg.com` the ability to steal credentials from every account.

Comment (grzuy) [reply app/views/layouts/application.html.erb]: Good call. Needs to change. Thank you :+1:

Comment (DanielHeath) [reply app/views/layouts/application.html.erb]: For that matter, the `secure.gaug.es` stuff stands out to me as pretty high-risk also. I'm sure the staff there are all very fine people, but they still shouldn't have access to my unhashed password.

File: config/locales/en.yml [hunk 4]
```diff
@@ -373,3 +374,13 @@ en:
         one: Displaying <b>1</b> %{model}
         other: Displaying <b>all&nbsp;%{count}</b> %{model}
         zero: No %{model} found
+  webauthn_credentials:
+    incorrect_credentials: Your credentials are incorrect.
+    index:
+      go_settings: Register a WebAuthn credential
```
Comment (bdewater) [inline config/locales/en.yml]: Major websites are going with "security keys" (e.g. [Google](https://support.google.com/accounts/answer/6103523), [Facebook](https://www.facebook.com/help/401566786855239), [Twitter](https://help.twitter.com/en/managing-your-account/two-factor-authentication), [Shopify](https://help.shopify.com/en/manual/your-account/account-security/two-step-authentication#enable-two-step-authentication-with-security-keys), [GitHub](https://github.blog/2019-08-21-github-supports-webauthn-for-security-keys/)), naming things differently might be confusing for users who are not familiar with the underlying specifications.

File: app/controllers/webauthn_credentials_controller.rb [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+    render json: { status: status, redirect_path: webauthn_credentials_path }, status: status
+  end
+
+  private
+
+  def require_webauthn_disabled
+    return unless current_user.webauthn_enabled?
+    flash[:error] = t(".require_webauthn_disabled")
+    render json: { status: :unprocessable_entity, redirect_path: webauthn_credentials_path }, status: status
```
Comment (bdewater) [inline app/controllers/webauthn_credentials_controller.rb]: I know it's mentioned as "what it doesn't" but I'm assuming you are planning to address this before general availability? 😅 Not allowing multiple security keys server no real purpose... Also see [New Device Registration](https://www.w3.org/TR/webauthn-1/#usecase-new-device-registration):
> Note: This approach of registering multiple authenticators for an account is also useful in account recovery use cases.

File: db/migrate/20190814131051_create_credentials.rb [hunk 6]
```diff
@@ -0,0 +1,11 @@
+class CreateCredentials < ActiveRecord::Migration[5.2]
+  def change
+    create_table :credentials do |t|
+      t.references :user, foreign_key: true
+      t.string :external_id, null: false
+      t.text :public_key, null: false
```
Comment (bdewater) [inline db/migrate/20190814131051_create_credentials.rb]: - There [SHOULD](https://www.w3.org/TR/webauthn-2/#sctn-sign-counter) be a `sign_count`.
- A name to help keep keys apart (assuming multiple keys will be allowed in the future)
- Also suggested on the W3C mailing list of GitHub repo as a best practice: a "last logged in at" timestamp to help disambiguate keys and potentially spot suspicious activity

File: app/controllers/sessions_controller.rb [hunk 7]
```diff
@@ -21,6 +24,48 @@ def mfa_create
     end
   end
 
+  def webauthn_authentication_options
+    user = User.find_by_name(session[:mfa_user])
```
Comment (eliotsykes) [inline app/controllers/sessions_controller.rb]: I see this is borrowing from existing code, which is understandable. That existing code and this would benefit from being specific about how the lookup is done and only lookup by handle as this is what is stored in `session[:mfa_user] = @user.handle`.

`User.find_by_name` searches by email first and only if that fails, does it lookup by handle:

```rb
def self.find_by_name(name)
  find_by(email: name) || find_by(handle: name)
end
```

Alternatively, store user id `session[:mfa_user] = @user.id` and lookup by `User.find(session[:mfa_user])`.

File: app/controllers/sessions_controller.rb [hunk 8]
```diff
@@ -21,6 +24,48 @@ def mfa_create
...
+      end
+    else
+      respond_to do |format|
+        format.json { render json: { errors: ["Unprocessable request"] }, status: :unprocessable_entity }
+      end
+    end
+  end
+
+  def webauthn_authentication
+    @user = User.find_by_name(session[:mfa_user])
```
Comment (eliotsykes) [inline app/controllers/sessions_controller.rb]: As above, avoid `User.find_by_name`, prefer a more specific find by handle or user id.

File: app/models/user.rb [hunk 9]
```diff
@@ -172,13 +174,22 @@ def mfa_enabled?
...
+
   def disable_mfa!
     mfa_disabled!
     self.mfa_seed = ""
     self.mfa_recovery_codes = []
     save!(validate: false)
   end
 
+  def disable_webauthn!
+    credentials.first.destroy!
```
Comment (eliotsykes) [inline app/models/user.rb]: Consider if this should `credentials.destroy_all`, or if the credentials association ought to be a `has_one` instead of `has_many`.

File: db/migrate/20190814131051_create_credentials.rb [hunk 10]
```diff
@@ -0,0 +1,11 @@
+class CreateCredentials < ActiveRecord::Migration[5.2]
+  def change
+    create_table :credentials do |t|
```
[the remark this answers is not in the document]
Comment (grzuy) [reply db/migrate/20190814131051_create_credentials.rb]: For sure not just `credentials`, I commented above maybe changing it to `webauthn_credentials`. I think I slightly prefer `webauthn_credentials` over `public_key_credentials`. I expect that to be more obvious for anyone playing with this codebase in the future...

File: config/locales/en.yml (hunk 4 above)
Comment (grzuy) [reply config/locales/en.yml]: Agree it might be better than "WebAuthn credential".

Although still possibly confusing for a user to map the "Security Key" concept to "I used my Android Fingerprint" or "MacBook TouchID".

In general, "Security Key" might better match the idea of a roaming authenticator. Not so much a platform authenticator I guess. I wonder if for the latter, the "Device" concept might be more intuitive.

File: app/controllers/webauthn_credentials_controller.rb (hunk 5 above)
Comment (grzuy) [reply app/controllers/webauthn_credentials_controller.rb]: Indeed.

That "What it doesn't" section sounds a bit confusing.
Updated and moved everything to "Pending" list.

File: app/models/user.rb (hunk 9 above)
Comment (grzuy) [reply app/models/user.rb]: In the end there should be multiple credentials so `has_many`.

And about destruction, it will probably make sense to have a delete for each of the credentials. Not sure we'll keep this `disable_webauthn!` method to delete all at once.

File: app/controllers/sessions_controller.rb (hunk 7 above)
Comment (grzuy) [reply app/controllers/sessions_controller.rb]: Good call.
Thank you!

File: app/controllers/webauthn_credentials_controller.rb (hunk 5 above)
Comment (bdewater) [reply app/controllers/webauthn_credentials_controller.rb]: Thanks!

File: config/locales/en.yml (hunk 4 above)
Comment (bdewater) [reply config/locales/en.yml]: Our original docs called it a 'security device' with similar underlying reasoning, we also tried to explain the difference between a USB key and a built-in biometric sensor. In the end we found these distinctions made it all just more confusing to read and we changed it to just 'security key' shortly thereafter.

Google is using the same naming everywhere, Chrome prompts me to choose "Use your security Key with Bluetooth/NFC/Fingerprint" on websites. I.e. a device can have a built-in security key or you can use an external one, but both are security keys.

Google also recently [announced that an Android phone's built-in security key can be used with a desktop or laptop](https://venturebeat.com/2019/04/10/you-can-now-use-your-android-phone-as-a-2fa-security-key-for-google-accounts/) (using the caBLE transport). So in technical terms; your phone's platform authenticator becomes also a roaming authenticator 🙈

Long story short, I'd suggest to not focus on the form factor or transport used but on the functionality: all of them are a security key with all the security properties you can expect from one 🙂

Comment (grzuy) [reply config/locales/en.yml]: Fair enough.

File: vendor/assets/javascripts/webauthn-json.js [hunk 11]
```diff
@@ -0,0 +1 @@
...required:!1,schema:e}}const s={type:o(r),id:o(i),transports:c(r)},u={publicKey:o({rp:o(r),user:o({id:o(i),name:o(r),displayName:o(r),icon:c(r)}),challenge:o(i),pubKeyCredParams:o(r),timeout:c(r),excludeCredentials:c([s]),authenticatorSelection:c(r),attestation:c(r),extensions:c(r)}),signal:c(r)},l={type:o(r),id:o(r),rawId:o(i),response:o({clientDataJSON:o(i),attestationObject:o(i)})},d={unmediated:c(r),mediation:c(r),publicKey:o({challenge:o(i),timeout:c(r),rpId:c(r),allowCredentials:c([s]),userVerification:c(r),extensions:c(r)}),signal:c(r)},f={type:o(r),id:o(r),rawId:o(i),response:o({clientDataJSON:o(i),authenticatorData:o(i),signature:o(i),userHandle:o(i)})},p={credentialCreationOptions:u,publicKeyCredentialWithAttestation:l,credentialRequestOptions:d,publicKeyCredentialWithAssertion:f};e.create=async function(e){const r=a(t,u,e),i=await navigator.credentials.create(r);return a(n,l,i)},e.get=async function(e){const r=a(t,d,e),i=await navigator.credentials.get(r);return a(n,f,i)},e.schema=p,e.supported=function(){return!!(navigator.credentials&&navigator.credentials.create&&navigator.credentials.get&&window.PublicKeyCredential)},Object.defineProperty(e,"__esModule",{value:!0})});
```
Comment (eliotsykes) [inline vendor/assets/javascripts/webauthn-json.js]: Avoid adding a hard to review file to the project, use an uncompressed version of `webauth-json.js` instead.

File: app/controllers/webauthn_credentials_controller.rb [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+      flash[:error] = t(".problem")
+      status = :unprocessable_entity
+    end
+
+    render json: { status: status, redirect_path: webauthn_credentials_path }, status: status
+  end
+
+  def destroy
+    begin
+      current_user.webauthn_credentials.find(params[:id]).destroy
```
Comment (eliotsykes) [inline app/controllers/webauthn_credentials_controller.rb]: `#destroy` can fail silently. Consider using `destroy!` or checking the return value of `destroy`.

File: config/routes.rb [hunk 13]
```diff
@@ -126,6 +126,13 @@
 
   ################################################################################
   # UI
+
+  resources :webauthn_credentials, only: %i[index create destroy] do
+    collection do
+      get :create_options
```
Comment (eliotsykes) [inline config/routes.rb]: `get :create_options` route is not protected against CSRF. Consider if this should use `post` instead to gain CSRF protection.

File: config/routes.rb [hunk 14]
```diff
@@ -169,6 +176,10 @@
 
     resource :session, only: %i[create destroy] do
       post 'mfa_create', to: 'sessions#mfa_create', as: :mfa_create
+      collection do
+        get :webauthn_authentication_options, constraints: { format: :json }, defaults: { format: :json }
```
Comment (eliotsykes) [inline config/routes.rb]: Consider altering the `:webauthn_authentication_options` route to use `post` instead of `get` to take advantage of Rails' CSRF protection.

**Comment by eliotsykes** [comment]
Would any of the new WebAuthn routes need throttling to mitigate brute force attacks?

File: app/controllers/internal/webauthn_registrations_controller.rb [hunk 15]
```diff
@@ -0,0 +1,45 @@
+class Internal::WebauthnRegistrationsController < ApplicationController
+  def options
+    current_user.update(webauthn_handle: bin_to_str(SecureRandom.random_bytes(64))) unless current_user.webauthn_handle
```
[the remark this answers is not in the document]
Comment (grzuy) [reply app/controllers/internal/webauthn_registrations_controller.rb]: Created issue: https://github.com/cedarcode/webauthn-ruby/issues/258

File: Dockerfile [hunk 16]
```diff
@@ -14,6 +14,7 @@ RUN apk add --no-cache \
   libxml2-dev \
   libxslt-dev \
   tzdata \
+  git \
```
Comment (grzuy) [inline Dockerfile]: Can be removed after we point the `webauthn` gem to a released version, if not wanted.

File: config/routes.rb (hunk 13 above)
Comment (grzuy) [reply config/routes.rb]: This action is not a state-changing action.

We renamed it, `create_options` was a bit confusing.

File: config/routes.rb (hunk 14 above)
Comment (grzuy) [reply config/routes.rb]: Was coded as a `GET` because not a state-changing action.

File: app/controllers/internal/webauthn_registrations_controller.rb (hunk 15 above)
[the remark this answers is not in the document]
Comment (grzuy) [reply app/controllers/internal/webauthn_registrations_controller.rb]: Done.

File: Dockerfile (hunk 16 above)
Comment (grzuy) [reply Dockerfile]: Removed.

File: vendor/assets/javascripts/webauthn-json.js [hunk 17]
```diff
@@ -0,0 +1,184 @@
+// webauthn-json v0.3.6
+//
+// Source: https://github.com/github/webauthn-json/tree/v0.3.1
+// License: MIT - https://github.com/github/webauthn-json/blob/v0.3.1/LICENSE.md
+//
+// Generated by checking out @github/webauthn-json source code and running
+// $ ROLLUP_WATCH=1 yarn build
```
Comment (grzuy) [inline vendor/assets/javascripts/webauthn-json.js]: @lgarron Hi, thanks for webauthn-json!

Had to run the above to include @github/webauthn-json here.
Would you recommend something different?

Comment (lgarron) [reply vendor/assets/javascripts/webauthn-json.js]: The *recommended* approach is to use `npm`, and we expect that most projects will use a build system.

If not, you can get a browser (UMD) build using:

    npm install @github/webauthn-json
    cat node_modules/@github/webauthn-json/dist/index.js

If you include this in a file on a page the old-school way, you should be able to access `webauthnJSON` (e.g. on `window`).

Your approach generates an equivalent un-minified file. If you want the un-minified file and don't care about code size, then that's probably the best way to do it.

Comment (grzuy) [reply vendor/assets/javascripts/webauthn-json.js]: Hi @lgarron,

Thanks for the response!
We're good for now this way then.

Comment (grzuy) [reply vendor/assets/javascripts/webauthn-json.js]: > The _recommended_ approach is to use `npm`, and we expect that most projects will use a build system.

Yep. Opened the following (https://github.com/rubygems/rubygems.org/issues/2126) while ago, just because of this.

File: app/views/sessions/mfa_prompt.html.erb [hunk 18]
```diff
@@ -0,0 +1,7 @@
+<% @title = t('multifactor_authentication') %>
+
+<% if @user&.webauthn_enabled? %>
+  <%= render "sessions/webauthn_prompt" %>
+<% end %>
+
+<%= render "sessions/otp_prompt" %>
```
Comment (sonalkr132) [inline app/views/sessions/mfa_prompt.html.erb:7]: Can we please this a horizontal split? vertical looks clunky. `l-half--l` and `l-half--r` may be helpful.

File: config/locales/de.yml [hunk 19]
```diff
@@ -16,6 +16,8 @@ de:
   please_sign_in:
   otp_incorrect:
   otp_missing:
+  use_key: Sign in with Security Key
```
Comment (sonalkr132) [inline config/locales/de.yml:19]: Just add keys (empty value) for non-english locales.

File: app/views/profiles/edit.html.erb [hunk 20]
```diff
@@ -102,6 +102,8 @@
       </div>
       <%= submit_tag t('.mfa.update'), :class => 'form__submit' %>
     <% end %>
+
+    <h2><%= link_to t('webauthn_credentials.index.title'), webauthn_credentials_path %></h2>
```
Comment (sonalkr132) [inline app/views/profiles/edit.html.erb:106]: As this stands, the user can only add webauth security keys after they have registered a TOTP authenticator device. I am not sure I understand this requirement. Shouldn't the user be able to enable mfa and only register webauth keys?

**Comment by sonalkr132** [comment]
Please add a section in CONTRIBUTING.md (or a new doc page) for steps to tests this locally. I was able to use https://github.com/google/virtual-authenticators-tab but it would nice to have this or perhaps an alternate method documented where we may not have to purchase a yubikey. 

We will also need guides (for users) similar to https://guides.rubygems.org/setting-up-multifactor-authentication/
perhaps clarifying the downside that this method doesn't support `gem` cli.

File: app/controllers/webauthn_credentials_controller.rb [hunk 21]
```diff
@@ -0,0 +1,22 @@
+class WebauthnCredentialsController < ApplicationController
+  before_action :redirect_to_signin, unless: :signed_in?
+
+  def index
+    @user = current_user
+  end
+
+  def destroy
```
Comment (sonalkr132) [inline app/controllers/webauthn_credentials_controller.rb:8]: Is it acceptable to allow users to delete the key without verifying that they have access to the key?

File: app/models/user.rb [hunk 22]
```diff
@@ -218,6 +224,26 @@ def otp_verified?(otp)
     save!(validate: false)
   end
 
+  def webauthn_verified?(current_challenge, webauthn_credential)
```
Comment (sonalkr132) [inline app/models/user.rb:227]: I am not sure I under why this method is in `User` model. Would it be better if we asked `WebauthnCredential` object if it was valid/verified? Do we want to ensure that webauthn_credential belongs to the user?

**Comment by sonalkr132** [comment]
Closing this in favour of #2865
